### DISCUSSION TITLE
Cancel the current drag-and-drop operation when pressing right click while dragging something

### DISCUSSION THREAD

**Discussion by rikmarais** [godotengine/godot-proposals#11048] [GUI]
Closely related to #4679

Currently when performing a drag operation, if I press a key mapped to `ui_cancel` it correctly cancels the drag as introduced in https://github.com/godotengine/godot/pull/68418

But if `ui_cancel` is mapped to a mouse button (e.g. right-click), it does not cancel the drag as I expected it would.
I looked into the cpp code and saw in https://github.com/godotengine/godot/pull/68418/files#diff-7d37141bff1285fa01a0530677bdd0e2d07f1a3d6a9644b7eb5555807b8445c0R1970-R1976 that `mb.is_null()` is one of the checks. 

I believe `mb` refers to the mouse button index associated with the action, meaning that a `ui_cancel` originating from a mouse button won't be able to hit the `_perform_drop` private method in this code.

The only similar discussion I could find on the topic was https://github.com/godotengine/godot-proposals/discussions/9044 which was implemented for the 4.3 milestone (apologies if unrelated, I'm still getting familiar) but you can't use `ui_cancel` on a mouse button to cancel a drag even in 4.3, so I don't think it affected the code in question.

Is there a particular reason why `mb.is_null()` needs to be true in order to cancel the drop?
Using right-click to cancel feels much more intuitive while mid-drag and my project involves dragging Controls (and sometimes cancelling the drag).

**Comment by Calinou** [comment]
I think it's just an oversight from when the code was initially written (note that the linked commit doesn't modify the check). That said, I'm not sure if remapping the right mouse button to `ui_cancel` is a good idea, as it'll affect all uses of `ui_cancel`.

We may want to introduce a dedicated built-in action for canceling drag-and-drop such as `ui_drag_and_drop_cancel`, and bind it to <kbd>Escape</kbd> and <kbd>Right mouse button</kbd> by default.

cc @Sauermann

**Reply by rikmarais** [reply]
Thanks for commenting! Aha, yes, that does make a lot of sense.

> I think it's just an oversight from when the code was initially written (note that the linked commit doesn't modify the check)

Very possible! I was curious if the block inside the `mb.is_null()` check was specifically targeted for a reason that I'm unfamiliar with.👍 

> That said, I'm not sure if remapping the right mouse button to ui_cancel is a good idea, as it'll affect all uses of ui_cancel.

Yeah, this use does indeed feel like it is more deliberately targeting a specific desired outcome than the general `ui_cancel`, but since the Viewport method to perform/cancel the drop isn't public outside it was the only way to do it at present.

A public viewport method to perform/cancel the drop or a dedicated `ui_drag_and_drop_cancel` built-in action (checked for outside of the block inside the check for `mb.is_null()`) seems that it would solve the problem handily, yes!

Edit: I'd be curious to see if others would consider this functionality useful. If so, I might try an issue/PR.

**Reply by Sauermann** [reply]
If I remember correctly, there is a design decision, that mouse buttons should not be used as actions.
The `mb.is_null()` check enforces this. So without revising that design decision, I don't see how to implement binding the right mouse button to an `ui_drag_and_drop_cancel` action.

See: https://github.com/godotengine/godot/pull/100259

**Comment by Sauermann** [comment]
This got implemented in https://github.com/godotengine/godot/pull/100259. I'm closing it accordingly.
